server/stream.go

```diff
@@ -931,18 +932,35 @@ func (mset *Stream) processInboundJetStreamMsg(_ *subscription, pc *client, subj
 	}
 
 	// If we are interest based retention and have no consumers then skip.
-	if interestRetention && numConsumers == 0 {
-		seq = store.SkipMsg()
-		if doAck && len(reply) > 0 {
-			response = append(pubAck, strconv.FormatUint(seq, 10)...)
-			response = append(response, '}')
-			mset.sendq <- &jsPubMsg{reply, _EMPTY_, _EMPTY_, nil, response, nil, 0}
+	if interestRetention {
+		var noInterest bool
+
+		if numConsumers == 0 {
+			noInterest = true
+		} else if mset.numFilter > 0 {
```

Comment:
We access mset.numFilter and mset.consumers without the mset lock here. Would recommend determining if there is noInterest for interestRetention early when still under the mset lock (when we determine the consumer count). It will also avoid differences between the time we evaluate the count and when we are here (since lock has been released number of consumers may have changed - up or down).